src/main/java/com/exasol/udfstartuptimeimprover/UdfStartUpTimeImproverInt.java

```diff
@@ -32,16 +36,66 @@
         final UdfDefinition udfDefinition = new UdfDefinitionParser().parseUdfDefinition(udfDefinitionString);
-        assertClassListExists(pathToClassListWithOffset);
-        runClassDumpGeneration(udfDefinition.getJars(), pathToClassListWithOffset);
+        final Path classListPath = extractClassList(udfDefinition);
+        runClassDumpGeneration(udfDefinition.getJars(), classListPath);
         uploadDump(bucket, pathForDump);
         return rewriteUdfDefinition(udfDefinition, bucket.getBucketFsName(), bucket.getBucketName(), pathForDump);
     }
 
+    private Path extractClassList(final UdfDefinition udfDefinition) {
+        final List<String> classLists = udfDefinition.getJars().stream().flatMap(this::findClassListInJar)
+                .collect(Collectors.toList());
+        if (classLists.isEmpty()) {
+            throw new IllegalStateException(ExaError.messageBuilder("E-USTI-17")
+                    .message("Non of the jars of this script contained a " + CLASSES_LIST_FILE_NAME + " file.")
+                    .mitigation(
+                            "Please check that the jar you're trying to optimize is \"Prepared for Java UDF startup time improver\".")
+                    .toString());
+        }
+        final String combinedClassList = combineClassLists(classLists);
+        return writeClassListToTmp(combinedClassList);
+    }
+
+    private Path writeClassListToTmp(final String combinedClassList) {
+        final Path classListPath = Path.of("/tmp/classes.lst");
+        try {
+            Files.writeString(classListPath, combinedClassList);
+        } catch (final IOException exception) {
+            throw new IllegalStateException(
+                    ExaError.messageBuilder("F-USTI-18")
+                            .message("Failed to write class list temporary to {{path}}.", classListPath).toString(),
+                    exception);
+        }
+        return classListPath;
+    }
+
+    private String combineClassLists(final List<String> classLists) {
+        return classLists.stream().flatMap(list -> Arrays.stream(list.split("\n")).map(String::trim))
+                .collect(Collectors.joining("\n"));
+    }
+
+    private Stream<String> findClassListInJar(final String jarName) {
+        final Path jarFile = Path.of(this.rootDirOffset + jarName);
+        try (final FileInputStream inputStream = new FileInputStream(jarFile.toFile());
+                final BufferedInputStream bufferedStream = new BufferedInputStream(inputStream);
+                final ZipInputStream zipInputStream = new ZipInputStream(bufferedStream)) {
+            ZipEntry entry = zipInputStream.getNextEntry();
+            while (entry != null) {
+                if (CLASSES_LIST_FILE_NAME.equals(entry.getName())) {
+                    return Stream.of(new String(zipInputStream.readAllBytes(), StandardCharsets.UTF_8));
+                }
+                entry = zipInputStream.getNextEntry();
+            }
```

Comment:
Instead of reading the complete JAR you could also use `ZipFile` (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/zip/ZipFile.html) which allows you to directly get the expected entry.

But I guess it won't make a big difference.